src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,766 @@
+    std::pair<instruction_ref, instruction_ref>
+    apply_broadcast_op(const onnx_parser::node_info& info,
+                       instruction_ref opl,
+                       instruction_ref opr,
+                       const int_vec& common_labels) const
+    {
+        std::pair<instruction_ref, instruction_ref> ret;
+
+        auto llens = opl->get_shape().lens();
+        auto rlens = opr->get_shape().lens();
+
+        bool lbc = false;
+        bool rbc = false;
+        for(auto l : common_labels)
+        {
+            if(llens[l] == 1 and rlens[l] == 1)
+                continue;
+
+            if(llens[l] == 1)
+            {
+                lbc      = true;
+                llens[l] = rlens[l];
+            }
+
+            if(rlens[l] == 1)
+            {
+                rbc      = true;
+                rlens[l] = llens[l];
+            }
+        }
+
+        if(lbc)
+            opl = info.add_instruction(make_op("multibroadcast", {{"out_lens", llens}}), opl);
+        if(rbc)
+            opr = info.add_instruction(make_op("multibroadcast", {{"out_lens", rlens}}), opr);
+
+        ret.first  = opl;
+        ret.second = opr;
+        return ret;
+    }
+
+    instruction_ref apply_reduce_sum_op(const onnx_parser::node_info& info,
+                                        instruction_ref op,
+                                        const int_vec& axes,
+                                        int_vec& row) const
+    {
+        if(axes.empty())
+            return op;
+
+        for(int a : axes)
+            row[a] = -1;
+
+        return info.add_instruction(make_op("reduce_sum", {{"axes", axes}}), op);
+    }
+
+    // Utility
+
+    int_mat make_matrix(int cur_pair, int cols, int fill_value) const
+    {
+        return int_mat(cur_pair, int_vec(cols, fill_value));
+    }
+
+    int_vec extract_column(int_mat map_mat, int col_idx, int row_begin, int row_end) const
+    {
+        int_vec ret;
+        ret.reserve(row_end - row_begin);
+
+        for(int i = row_begin; i < row_end; ++i)
+            ret.push_back(map_mat[i][col_idx]);
+
+        return ret;
+    }
+
+    int_vec set_union(const int_vec& lhs, const int_vec& rhs) const
+    {
+        int_vec ret;
+        std::set_union(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), std::back_inserter(ret));
+
+        return ret;
+    }
+
+    int_vec set_difference(const int_vec& lhs, const int_vec& rhs) const
+    {
+        int_vec ret;
+        std::set_difference(
+            lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), std::back_inserter(ret));
+
+        return ret;
+    }
+
+    // Equivalent to numpy.arange without the step parameter
+    int_vec arange(int start_value, int end_value) const
+    {
+        int_vec ret(end_value - start_value);
+        std::iota(ret.begin(), ret.end(), start_value);
+        return ret;
+    }
+
+    template <typename... Vecs>
+    std::decay_t<std::tuple_element_t<0, std::tuple<Vecs...>>> concat_vectors(Vecs&&... vecs) const
```

Comment:
Take the first vector as a seperate parameter instead of using `tuple_element`:

```cpp
template<class Vector, class... Vectors>
Vector concat_vectors(Vector vec, Vectors&&... vecs)
{
    ...
    return vec;
}
```

Reply:
Done.